### ISSUE TITLE
data_load_ladscape是否可以上传呢

### ISSUE THREAD

**Issue by ech047** [QUSL0414/QUSL#1]

**Comment by UESTC-YLH** [comment]
import os
import time
import random
import numpy as np
import pandas as pd
from PIL import Image
from os import listdir
import cg_genetic_algorithm
from sklearn.svm import SVC
from datetime import datetime
import matplotlib.pyplot as plt
from functools import partial
from multiprocessing import Pool
from scipy.stats import spearmanr
from qiskit import QuantumCircuit
from os.path import isfile, join
from sklearn import preprocessing
from sklearn.decomposition import PCA
from concurrent.futures import ThreadPoolExecutor
from projectq.ops import (H, X, Y, Z, T, Tdagger, S, Sdagger, CNOT, CX, Rx,
                          Ry, Rz, SqrtX, get_inverse, Swap, SwapGate)

def generate_landscape_triplets(base_path, dataset, num_triplets, testing=False):
    base_path = "/home/ubuntu/mnist/"
    path = f'{base_path}//DATASET/{dataset}/{dataset}_train'
    path_test = f'{base_path}//DATASET/{dataset}/{dataset}_test'
    image_triplets = []
    image_indecies = []
    if testing:
        files = [f for f in listdir(path_test) if isfile(join(path_test, f))]
        files.sort()
        num_files = len(files)
        # new_width, new_height = 18, 18
        new_width, new_height = 32, 32
        for _ in range(num_triplets):
            indecies = []
            triplets = []
            images = []
            for i in range(3):
                index = int(np.floor(random.random() * num_files))
                im = Image.open(f'{base_path}//DATASET/{dataset}/{dataset}_test/' + str(files[index]))

                width, height = im.size
                left = (width - new_width) // 2
                top = (height - new_height) // 2
                right = (width + new_width) // 2
                bottom = (height + new_height) // 2
                im = im.crop((left, top, right, bottom))
                im = im.resize((9, 9), Image.Resampling.LANCZOS)
                r_bin, g_bin, b_bin = get_color_density(np.array(im))

                triplets.append(np.array(r_bin + g_bin + b_bin))
                indecies.append(index)

                images.append(np.reshape(np.array(im), [-1, ]))

            noisy_image = add_noise(np.array(images[0]), noise_type='gaussian', seed=42)

            image_triplets.append((images[0], noisy_image, images[2]))
            image_indecies.append((indecies[0], indecies[0], indecies[2]))

    else:
        files = [f for f in listdir(path) if isfile(join(path, f))]
        files.sort()
        num_files = len(files)
        new_width, new_height = 32, 32

        for _ in range(num_triplets):
            indecies = []
            triplets = []
            images = []
            for i in range(3):
                index = int(np.floor(random.random() * num_files))
                im = Image.open(f'{base_path}/DATASET/{dataset}/{dataset}_train/' + str(files[index]))


                width, height = im.size
                left = (width - new_width) // 2
                top = (height - new_height) // 2
                right = (width + new_width) // 2
                bottom = (height + new_height) // 2
                im = im.crop((left, top, right, bottom))
                im = im.resize((9, 9), Image.Resampling.LANCZOS)

                r_bin, g_bin, b_bin = get_color_density(np.array(im))
                triplets.append(np.array(r_bin + g_bin + b_bin))
                indecies.append(index)
                images.append(np.reshape(np.array(im), [-1, ]))

            noisy_image = add_noise(np.array(images[0]), noise_type='gaussian', seed=42)
            image_triplets.append((images[0], noisy_image, images[2]))
            image_indecies.append((indecies[0], indecies[0], indecies[2]))
    return image_triplets, image_indecies


def add_noise(image, noise_type='gaussian', seed=None):
    """
    Add noise to an image.

    Parameters:
        image (numpy.ndarray): Input image.
        noise_type (str): Type of noise ('gaussian', 'salt_and_pepper', 'poisson').
        seed (int): Seed for random number generation.

    Returns:
        numpy.ndarray: Image with added noise.
    """
    if seed is not None:
        np.random.seed(seed)

    noisy_image = image.copy()

    if noise_type == 'gaussian':
        mean = 0
        std = 5  # Adjust the standard deviation based on your preference
        noise = np.random.normal(mean, std, image.shape)
        noisy_image = np.clip(image + noise, 0, 255)

    elif noise_type == 'salt_and_pepper':
        prob = 0.05  # Adjust the probability based on your preference
        mask = np.random.rand(*image.shape) < prob
        noisy_image[mask] = 0 if np.random.rand() < 0.5 else 255

    elif noise_type == 'poisson':
        noisy_image = np.random.poisson(image)

    return noisy_image.astype(np.uint8)


def perform_pca(x, pca_dims=4096):
    pca = PCA(pca_dims)
    # Normalize image data so its pythagorean sum is 1
    pca.fit(preprocessing.normalize(x))
    return pca.transform(preprocessing.normalize(x))


def get_color_density(image):
    num_bins = 8
    bin_threshold = 256 // num_bins
    red = image[..., 0]
    green = image[..., 1]
    blue = image[..., 2]
    red_bin = np.reshape(red // bin_threshold, [-1, ])
    green_bin = np.reshape(green // bin_threshold, [-1])
    blue_bin = np.reshape(blue // bin_threshold, [-1])
    red_count = [0 for _ in range(num_bins)]
    blue_count = [0 for _ in range(num_bins)]
    green_count = [0 for _ in range(num_bins)]
    for r, g, b, in zip(red_bin, blue_bin, green_bin):
        red_count[r] += 1
        blue_count[g] += 1
        green_count[b] += 1
    return red_count, blue_count, green_count


def crop_and_resize_image(image):
    new_width, new_height = 32, 32
    width, height = image.size
    left = (width - new_width) // 2
    top = (height - new_height) // 2

    right = (width + new_width) // 2
    bottom = (height + new_height) // 2
    return image.crop((left, top, right, bottom))


def cosine_similarity(image1, image2):
    mean1 = np.mean(image1)
    mean2 = np.mean(image2)
    std1 = np.std(image1)
    std2 = np.std(image2)
    image1 = (image1 - mean1) / std1
    image2 = (image2 - mean2) / std2
    dot_product = np.dot(image1, image2)
    norm1 = np.linalg.norm(image1)
    norm2 = np.linalg.norm(image2)
    cosine_similarity = dot_product / (norm1 * norm2)
    return cosine_similarity


def process_feature(triplets, indecies, evolution, circuit, number_of_qubits, path, files, p, q, dataset, base_path):
    l = random.randint(0, 99)
    dists = []
    loss_tuples = []
    distance_tuples = []
    min_loss = 10e5
    closest_dist = 10e5
    closest_index = 0
    anchor_image = triplets[l][0]
    anchor_index = indecies[l][0]
    all_losses = []
    all_distances = []
    for image, indexes in zip(triplets, indecies):
        if indexes[0] != anchor_index:
            flattened = []
            for i, j, in zip(anchor_image, image[0]):
                flattened.append(i)
                flattened.append(j)
            loc1 = cg_genetic_algorithm.runcircuit(circuit, flattened, number_of_qubits=number_of_qubits)
            flattened = []
            for i, j, in zip(anchor_image, image[0]):
                flattened.append(j)
                flattened.append(i)
            loc2 = cg_genetic_algorithm.runcircuit(circuit, flattened, number_of_qubits=number_of_qubits)
            loss = sum([np.abs(i - j) for i, j in zip(loc1, loc2)])

            im1 = Image.open(join(path, str(files[anchor_index])))
            im1 = crop_and_resize_image(im1)
            im2 = Image.open(join(path, str(files[indexes[0]])))
            im2 = crop_and_resize_image(im2)

            anchor_r, anchor_g, anchor_b = get_color_density(np.array(im1))
            new_r, new_g, new_b = get_color_density(np.array(im2))
            dist = np.sum(
                np.power(np.array(anchor_r + anchor_g + anchor_b) - np.array(new_r + new_g + new_b), 2)) ** 0.5
            dists.append(dist)
            distance_tuples.append((float(dist), files[indexes[0]]))
            loss_tuples.append((float(loss), files[indexes[0]]))

            if loss < min_loss:
                min_loss = loss
                closest_index = indexes[0]
                closest_dist = dist

    all_losses.append(loss_tuples)
    all_distances.append(distance_tuples)
    loss_df = pd.DataFrame(loss_tuples, columns=['Loss', 'File'])
    distance_df = pd.DataFrame(distance_tuples, columns=['Distance', 'File'])

    loss_df.to_csv(f'{base_path}/result/{dataset}/csv/loss{q}_{p}.csv', index=False)
    distance_df.to_csv(f'{base_path}/result/{dataset}/csv/distance{q}_{p}.csv',
                       index=False)

    merged_df = pd.concat([loss_df, distance_df], axis=1)
    merged_df.to_csv(f'{base_path}/result/{dataset}/csv/distance_loss_merged_{q}_{p}.csv', index=False)

    loss_column = loss_df['Loss']
    distance_column = distance_df['Distance']
    merged_df = pd.concat([loss_column, distance_column], axis=1)

    merged_df.to_csv(f'{base_path}/result/{dataset}/csv/merged_data_{q}_{p}.csv', index=False)

    im1 = Image.open(join(path, str(files[anchor_index])))
    im1 = crop_and_resize_image(im1)
    im2 = Image.open(join(path, str(files[closest_index])))
    im2 = crop_and_resize_image(im2)
    print(
        f'anchor_index： {anchor_index}\nmin_loss： {min_loss}\nclosest_index： {closest_index}\nclosest_dist： {closest_dist}')
    Image.fromarray(np.hstack((np.array(im1), np.array(im2)))).save(
        f'{base_path}/result/{dataset}/SliqImages/{anchor_index}_{closest_index}_{q}_{p}.jpg')


def find_closest(base_path, dataset, number_of_qubits, i, j, triplets, indecies):
    start = time.perf_counter()
    individual_path = f'{base_path}result/{dataset}/qasm/{dataset}_next_generation_{j}_{i}.qasm'
    circuit = QuantumCircuit.from_qasm_file(individual_path)
    path = f'{base_path}//DATASET/{dataset}/{dataset}_test'
    files = [f for f in listdir(path) if isfile(join(path, f))]
    files.sort()
    evolution = cg_genetic_algorithm.Evolution(triplets)

    with ThreadPoolExecutor(max_workers=32) as executor:
        futures = []
        for _ in range(1):
            future = executor.submit(process_feature, triplets, indecies, evolution, circuit,
                                     number_of_qubits, path, files, i, j, dataset,base_path)
            futures.append(future)

        for future in futures:
            future.result()

    runtime2 = round(time.perf_counter() - start, 2)
    print("runtime test", runtime2)


def median_spearman_corr(x, y):
    x_rank = x.rank()
    y_rank = y.rank()
    spearman_corr, _ = spearmanr(x_rank, y_rank)
    return spearman_corr


def cnot(base_path, dataset, i, j):
    individual_path = f'{base_path}result/{dataset}/qasm/{dataset}_next_generation_{j}_{i}.qasm'
    circuit = QuantumCircuit.from_qasm_file(individual_path)
    produce_circuit = cg_genetic_algorithm.from_qiskit_circuit(circuit)
    cnot_count = 0
    for gate_info in produce_circuit:
        if gate_info[0] == "TFG" and gate_info[1] == CX:
            cnot_count += 1
    return cnot_count


def Spearman(base_path, dataset, i, j):
    print(f'###################{i}###################')
    csv_filename = f'{base_path}result/{dataset}/csv/merged_data_{j}_{i}.csv'
    df = pd.read_csv(csv_filename)

    all_losses = df['Loss']
    all_distances = df['Distance']
    plt.figure()
    plt.scatter(all_losses, all_distances)
    plt.title('Scatter Plot of Losses and Distances')
    plt.xlabel('Loss')
    plt.ylabel('Distance')
    plt.grid(True)

    median_spearman_corr_value = median_spearman_corr(all_losses, all_distances)

    print(f"Median Spearman correlation coefficient: {median_spearman_corr_value}")

    save_path = f'{base_path}result/{dataset}/correlation/plot_{j}_{i}_{median_spearman_corr_value}.png'
    plt.savefig(save_path)

    # plt.show()
    print(f'num{j}_{i}individual：{median_spearman_corr_value}')
    return median_spearman_corr_value
